### PR TITLE
[REVIEW] Coalescing dev/test/doc conda & pip envs, improving Dockerfile

### PR DESCRIPTION
**Pull request by randerzander**
For those interested, a brief history of the cuDF Dockerfile:

When we first created the Dockerfile, it was intended to make it easier for users to build libgdf and pygdf in one go. The dev_py35.yml conda environment included the latest master build of libgdf, meaning you couldn't use it to test your own local libgdf build.

Since then, dev_py35.yml was simplified, now only containing the deps needed to build cudf/cpp and cudf/python.

Lately there have been multiple Dockerfile issues dealing with conda package versions as specified directly in the Dockerfile. Instead of continuing to track and update versions in both the conda env yml file AND the Dockerfile, this PR sets up the Dockerfile to use the existing dev_py35.yml dep specifications.

In summary, this PR:
1. Moves conda deps from Dockerfile into conda/environments/dev_py35.yml
2. Eliminates conda package version args from the Dockerfile
3. Moves pip deps from docs/requirement.txt into conda/environments/dev_py35.yml
4. Sets up the Dockerfile to `ADD` necessary source files from local directories, instead of git cloning. This will let devs change cpp or python locally and build and test in the container without needing to first push to a public GitHub branch.
5. Bump numba dep version after comments from @kkraus14 
6. Fixes the Dockerfile so that pytests can be run without failure

@kkraus14 @mike-wendt who may opinions on item #4

### CODE REVIEW

File: conda/environments/dev_py35.yml [hunk 1]
```diff
@@ -18,3 +19,12 @@ dependencies:
 - cffi>=1.10.0   
 - distributed>=1.23.0
 - cython=0.28.*
+- pytest
+- sphinx
+- sphinx_rtd_theme
+- sphinxcontrib-websupport
+- numpydoc
+- ipython
+- pip:
```
Comment (kkraus14) [inline conda/environments/dev_py35.yml]:
```suggestion
- recommonmark
```

File: conda/environments/dev_py35.yml [hunk 2]
```diff
@@ -18,3 +19,12 @@ dependencies:
...
+  - recommonmark
```
Comment (kkraus14) [inline conda/environments/dev_py35.yml]:
```suggestion
  - pip:
```

File: conda/environments/dev_py35.yml [hunk 3]
```diff
@@ -18,3 +19,12 @@ dependencies:
...
 - distributed>=1.23.0
 - cython=0.28.*
+- pytest
+- sphinx
+- sphinx_rtd_theme
+- sphinxcontrib-websupport
+- numpydoc
+- ipython
+- recommonmark
+  - pip:
```
Comment (kkraus14) [inline conda/environments/dev_py35.yml]:
```suggestion
- pip:
```
